Add trpc guide (#7664)
edgedb/edgedb: docs/guides/tutorials/trpc.rst

.. _ref_guide_trpc:

Integrating EdgeDB with tRPC
============================

This guide explains how to integrate **EdgeDB** with **tRPC** for a modern, 
type-safe API. We'll cover setting up database interactions, API routing, 
and implementing authentication, all while ensuring type safety across the 
client and server.

You can reference the following repositories for more context:

> - `TurboRepo Template <https://github.com/scotttrinh/turbo-edb-app/blob/main-template/>`_: A monorepo template using **TurboRepo**, **EdgeDB**, and **tRPC**.

Reviewer: Maybe we should move or rebuild this into our official examples repo?

Author: Or maybe change it to an official template so that people can click "use it as a template"?

Reviewer: Yeah, maybe both? Can a subdirectory be a template? Probably not. I certainly don't mind linking to a personal repo of mine, but it feels like it's worth trying to make it a bit more official if we can. I guess since we already have the next+edb-auth template, there is already precedent for having a template repo in the edgedb org. I'll see about transferring it as-is and making it a template.

- `LookFeel Project <https://github.com/LewTrn/lookfeel>`_: A real-world project example using **EdgeDB** and **tRPC**.

Step 1: EdgeDB setup
====================

EdgeDB will serve as the database layer for your application.

Install and initialize EdgeDB
-----------------------------

To initialize **EdgeDB**, run the following command using your preferred 
package manager:

.. code-block:: bash

   $ pnpm dlx edgedb project init # or npx edgedb project init

This will create an EdgeDB project and set up a schema to start with.

Define the EdgeDB Schema
------------------------

The previous command generated a schema file in the `dbschema` directory.

Here’s an example schema that defines a `User` model:

> .. code-block:: sdl

Reviewer: Maybe add `:caption:` blocks to these files to make it a little more concrete what the structure is?

```suggestion
.. code-block:: sdl
   :caption: dbschema/default.esdl
```

   $ # or bun add edgedb

Then, create a client instance in a `edgedb.ts` file:

.. code-block:: typescript

   import { createClient } from 'edgedb';

   const edgedbClient = createClient();
   export default edgedbClient;

This client will be used to interact with the database and execute queries.

Step 3: tRPC setup
==================

**tRPC** enables type-safe communication between the frontend and 
backend.

Install tRPC dependencies
-------------------------

Install the required tRPC dependencies:

.. code-block:: bash

   $ pnpm add @trpc/server @trpc/client
   $ # or yarn add @trpc/server @trpc/client
   $ # or npm install @trpc/server @trpc/client
   $ # or bun add @trpc/server @trpc/client

If you're using React and would like to use React Query with tRPC, also 
install a wrapper around the `@tanstack/react-query <https://tanstack.com/query/latest>`_.

.. code-block:: bash

   $ pnpm add @trpc/react-query
   $ # or yarn add @trpc/react-query
   $ # or npm install @trpc/react-query
   $ # or bun add @trpc/react-query

Define the tRPC Router
-----------------------

Here’s how to define a simple tRPC query that interacts with **EdgeDB**:

.. code-block:: typescript

>    :caption: server/routers/_app.ts

Reviewer: Ahh, yeah, exactly.

Step 4: Use tRPC Client
========================

Now that the server is set up, you can use the tRPC client to interact with
the API from the frontend. We will demonstrate how to integrate tRPC with
**Next.js** and **Express**.

With Next.js
------------

If you're working with **Next.js**, here’s how to integrate **tRPC**:

Create a tRPC API Handler
~~~~~~~~~~~~~~~~~~~~~~~~~

Inside `api/trpc/[trpc].ts`, create the following handler to connect 
**tRPC** with Next.js:

.. code-block:: typescript
   :caption: pages/api/trpc/[trpc].ts

   import { createNextApiHandler } from '@trpc/server/adapters/next';
   import { appRouter } from '../../../server/routers/_app';

   export default createNextApiHandler({
     router: appRouter,
   });

Create a tRPC Client
~~~~~~~~~~~~~~~~~~~~

Next, create a **tRPC** client to interact with the API:

.. code-block:: typescript
   :caption: utils/trpc.ts

   import { createTRPCReact } from "@trpc/react-query";
   import { AppRouter } from './routers/_app';

   export const api = createTRPCReact<AppRouter>();

Client-Side Usage in Next.js
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can then use **tRPC** hooks to query the API from the client:

> .. code-block:: typescript

Reviewer: Is this a client component?

Author: It could be both

client-side tRPC calls can access the user’s session.

EdgeDB Auth with tRPC and tRPC context in Next.js
-------------------------------------------------

In **Next.js**, integrating **EdgeDB Auth** with **tRPC** involves creating a 
context that provides the user session and EdgeDB client to the tRPC API.

1. **Initialize EdgeDB Client and Auth**

   First, initialize the **EdgeDB** client and **EdgeDB Auth**:

   .. code-block:: typescript

      import { createClient } from "edgedb";
      import createAuth from "@edgedb/auth-nextjs/app";

      // Initialize EdgeDB client
      export const edgedbClient = createClient();

      // Initialize EdgeDB Auth
      export const auth = createAuth(edgedbClient, {
        baseUrl: process.env.VERCEL_ENV === "production"
          ? "https://production.yourapp.com"
          : "http://localhost:3000",
      });

2. **Create tRPC Context**

   The **tRPC** context provides the session and the EdgeDB client to the tRPC 
   procedures:

   .. code-block:: typescript

      import { initTRPC } from '@trpc/server';
      import { headers } from "next/headers";
      import { auth } from "./path-to-auth";

>       import { edgedbClient } from "./path-to-edgedb";

Reviewer: These are coming from the same module, I think, so if we make the module location and name explicit above, we can make this more explicit.

      // Create tRPC context with session and EdgeDB client
      export const createTRPCContext = async () => {
        const session = await auth.getSession(); // Retrieve session from EdgeDB Auth

        return {
          session, // Pass the session to the context
          edgedbClient, // EdgeDB client for querying the database
        };
      };

      // Initialize tRPC with context
      const t = initTRPC.context<typeof createTRPCContext>().create({});

3. **Use tRPC Context in API Handler**

   In **Next.js**, set up an API handler to connect your **tRPC router** with 
   the context:

   .. code-block:: typescript

      import { createNextApiHandler } from '@trpc/server/adapters/next';
      import { appRouter } from './path-to-router';

      export default createNextApiHandler({
        router: appRouter, // Your tRPC router
        createContext: createTRPCContext,
      });

4. **Example tRPC Procedure**

   You can now write procedures in your tRPC router, making use of the 
   **EdgeDB Auth** session and the **EdgeDB** client:

   .. code-block:: typescript

      export const appRouter = t.router({
        getUserData: t.procedure.query(async ({ ctx }) => {

>           if (!ctx.session) {
>             throw new Error("Not authenticated");
>           }
>           // Fetch data from EdgeDB using the client's session
>           const userData = await ctx.edgedbClient.query(`
>             SELECT User { name, email } WHERE .id = <uuid>$id
>           `, { id: ctx.session.user.id });

Reviewer: ```suggestion
          if (!(await ctx.session.isSignedIn())) {
            throw new Error("Not authenticated");
          }
          // Fetch data from EdgeDB using the authenticated client
          const userData = await ctx.session.client.query(`
            select User { name, email }
          `);
```

This assumes you've set up some access policies to disallow selecting users other than yourself. Otherwise:

```suggestion
          if (!(await ctx.session.isSignedIn())) {
            throw new Error("Not authenticated");
          }
          // Fetch data from EdgeDB using the authenticated client
          const userData = await ctx.session.client.query(`
            select User { name, email } filter global ext::auth::ClientTokenIdentity in .identities
          `);
```

(or for single identity<->user models `global ext::auth::ClientTokenIdentity = .identity`)

        })
      );

      app.listen(4000, () => {
        console.log('Server running on port 4000');
      });

4. **Example tRPC Procedure in Express**

   Once the context is set, you can define tRPC procedures that use both the 
   session and EdgeDB client:

   .. code-block:: typescript

      export const appRouter = t.router({
        getUserData: t.procedure.query(async ({ ctx }) => {
          if (!ctx.session) {
            throw new Error("Not authenticated");
          }
          // Query EdgeDB using the client's session
          const userData = await ctx.edgedbClient.query(`
            SELECT User { name, email } WHERE .id = <uuid>$id
          `, { id: ctx.session.user.id });

          return userData;
        }),
      });

Conclusion
----------

By integrating **EdgeDB Auth** into the tRPC context, you ensure that 
authenticated sessions are securely passed to API procedures, enabling 
user authentication and protecting routes.

You can also reference these projects for further examples:

- `TurboRepo Template <https://github.com/scotttrinh/turbo-edb-app/blob/main-template/>`_
- `LookFeel Project <https://github.com/LewTrn/lookfeel>`_ - A real-world 

>   example using **EdgeDB** and **tRPC**.

Reviewer: Do we want to restate these links here, or just keep them in the beginning or ending?

Author: I guess both options make sense: in the beginning for people who'd like to skip to the code, and at the end for a summary.

          createContext: (req) => req.context, // Use context from middleware
        })
      );

      app.listen(4000, () => {
        console.log('Server running on port 4000');
      });

4. **Example tRPC Procedure in Express**

   Once the context is set, you can define tRPC procedures that use both the 
   session and EdgeDB client:

   .. code-block:: typescript

      export const appRouter = t.router({
        getUserData: t.procedure.query(async ({ ctx }) => {
          if (!(await ctx.session.isSignedIn())) {
            throw new Error("Not authenticated");
          }
          // Fetch data from EdgeDB using the authenticated client
          const userData = await ctx.session.client.query(`
            select User { name, email }
          `);

          return userData;
        }),
      });

Conclusion
----------

By integrating **EdgeDB Auth** into the tRPC context, you ensure that 
authenticated sessions are securely passed to API procedures, enabling 
user authentication and protecting routes.

You can also reference these projects for further examples:

- `create-t3-turbo-edgedb <https://github.com/edgedb/create-t3-turbo-edgedb>`_ - 

>   A monorepo template using TurboRepo, EdgeDB, and tRPC.

Author: ```suggestion
  A monorepo template using Turborepo, EdgeDB, and tRPC.
```
